File: learn/src/main/java/org/platanios/learn/neural/training/LossFunction.java
package org.platanios.learn.neural.training;

import org.platanios.learn.math.matrix.Vector;
import org.platanios.learn.neural.network.Network;
import org.platanios.learn.neural.network.State;
import org.platanios.learn.neural.network.Variable;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Emmanouil Antonios Platanios
 */
public abstract class LossFunction {
    protected final Network network;

    public LossFunction(Network network) {
        this.network = network;
    }

    public abstract double value(State state, Vector correctOutput);
    public abstract Vector gradient(State state, Variable variable, Vector correctOutput);

    public List<Vector> gradient(State state, List<Variable> variables, Vector correctOutput) {
        return variables.stream()
                .map(variable -> gradient(state, variable, correctOutput))
                .collect(Collectors.toList());
    }
}

File: learn/src/main/scala/makina/scala/learn/neural/network/LossFunction.scala
package makina.scala.learn.neural.network

import breeze.linalg.{DenseVector, norm, sum}
import breeze.numerics.{log, pow}

/**
  * @author Emmanouil Antonios Platanios
  */
abstract class LossFunction {
  val trueOutput: DenseVector[Double]

  val inputSize: Int = trueOutput.size

  def value(networkOutput: DenseVector[Double]): Double
  def gradient(networkOutput: DenseVector[Double]): DenseVector[Double]
}

case class CrossEntropyLossFunction(trueOutput: DenseVector[Double]) extends LossFunction {
  override def value(networkOutput: DenseVector[Double]): Double =
    sum((log(networkOutput) :* trueOutput * -1.0) - log(networkOutput.map(1 - _) :* trueOutput.map(1 - _)))

  override def gradient(networkOutput: DenseVector[Double]): DenseVector[Double] =
    (trueOutput.map(1 - _) :/ networkOutput.map(1 - _)) - (trueOutput :/ networkOutput)
}

case class DifferenceL2NormSquaredLossFunction(trueOutput: DenseVector[Double]) extends LossFunction {
  override def value(networkOutput: DenseVector[Double]): Double = pow(norm(networkOutput - trueOutput), 2)
  override def gradient(networkOutput: DenseVector[Double]): DenseVector[Double] = (networkOutput - trueOutput) * 2.0
}
